### PR TITLE
Speed up reading uniqState

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Speed up reading UniquesHashSet (uniqState from disk for example)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

In some cases, the uniqMerge process for a query was completely dominated by the many calls to read data from the buffer into a uniqState state. Let's read all the integers in a block to avoid the overhead.

Example:

* Setup:
```sql
select version();

drop database if exists tt;

create database tt;

create table tt.landing (
	t DateTime,
	a String,
	b String
) Engine=MergeTree partition by tuple() order by a settings index_granularity = 1024;

create table tt.matview (
	a String,
	b_count AggregateFunction(uniq, String)
) Engine=AggregatingMergeTree partition by tuple() order by a settings index_granularity = 1024;

CREATE MATERIALIZED VIEW tt.mv TO tt.matview AS (
	SELECT a, uniqState(b) b_count
	FROM tt.landing
	GROUP BY a
);

insert into tt.landing select today()-1, toString(rand()%1000), concat('usr',toString(number%500)) from numbers(10000000);
insert into tt.landing select today()-2, toString(rand()%1000), concat('usr',toString(number%400)) from numbers(10000000);
insert into tt.landing select today()-3, toString(rand()%1000), concat('usr',toString(number%200)) from numbers(10000000);
```

* `clickhouse-benchmark --query "select a, uniqMerge(b_count) as b_count from tt.matview prewhere a='555' group by a FORMAT Null SETTINGS max_threads=1;" -i 2000`
  - Master: localhost:9000, queries 2000, QPS: 151.412, RPS: 151411.945, MiB/s: 21.207, result RPS: 0.000, result MiB/s: 0.000.
  - Changes: localhost:9000, queries 2000, QPS: 282.340, RPS: 282339.980, MiB/s: 39.545, result RPS: 0.000, result MiB/s: 0.000.

* `clickhouse-benchmark --query "select a, uniqMerge(b_count) as b_count from tt.matview group by a FORMAT Null SETTINGS max_threads=1;" -i 1000`
  - Master: localhost:9000, queries 1000, QPS: 81.189, RPS: 81188.559, MiB/s: 11.371, result RPS: 0.000, result MiB/s: 0.000.
  - Changes: localhost:9000, queries 1000, QPS: 105.919, RPS: 105919.390, MiB/s: 14.835, result RPS: 0.000, result MiB/s: 0.000.

### CODE REVIEW

**Comment by Algunenano** [comment]
Thinking out loud late at night: I wonder if this needs to handle endianness changes (where the generator of the data and the reader have different endianness). I haven't seen anything like that anywhere so I'd assume it's not supported in general, but I haven't looked much yet to confirm that's the case

**Comment by Algunenano** [comment]
Replying to myself: No other part of the code seems to consider endianness changes anywhere and there are similar approaches in statuses like `RoaringBitmapWithSmallSet` or `QuantileTimingTiny`. Also, there isn't any conversion in ReadHelpers.h (`readIntBinary` et. al) which would be the place where I'd add it.

So unless somebody says the opposite I'd say it's safe to assume you can't use data (via storage or network) generated by CH servers with a different endianness as the current server.

**Comment by Algunenano** [comment]
Looking at test errors:
- [Stateless tests (debug, s3 storage) [1/3]](https://s3.amazonaws.com/clickhouse-test-reports/41089/eae48c273b5bb46a350549a00a4fb10b08697ecd/stateless_tests__debug__s3_storage__[1/3].html) -> 00632_get_sample_block_cache. A timeout, likely related to s3 connection failing?
- [Stateless tests (release, s3 storage)](https://s3.amazonaws.com/clickhouse-test-reports/41089/eae48c273b5bb46a350549a00a4fb10b08697ecd/stateless_tests__release__s3_storage_.html) -> 01396_inactive_replica_cleanup_nodes_zookeeper. Seems like a s3 error: `AWSErrorMarshaller: Encountered AWSError 'NoSuchKey': The specified key does not exist.`
- [Stateless tests (tsan) [3/3]](https://s3.amazonaws.com/clickhouse-test-reports/41089/eae48c273b5bb46a350549a00a4fb10b08697ecd/stateless_tests__tsan__[3/3].html) -> `Server does not respond to health check`. The server took too long to reply so the test handler thought it was dead (but it wasn't) and stopped testing. No idea why the query (`4c155115-283b-48fb-9400-90cee8a267ac` `SELECT 'Running test stateless/01698_fix_toMinute.sql from pid=940'` took almost 2 minutes to finish).

File: src/AggregateFunctions/UniquesHashSet.h [hunk 1]
```diff
@@ -424,14 +424,30 @@ class UniquesHashSet : private HashTableAllocatorWithStackMemory<(1ULL << UNIQUE
...
+                else
+                    reinsertImpl(x);
+            }
+        }
+        else
+        {
+            auto hs = std::make_unique<HashValue[]>(m_size);
+            rb.readStrict(reinterpret_cast<char *>(hs.get()), m_size * sizeof(HashValue));
+
+            for (size_t i = 0; i < m_size; ++i)
```
Comment (nickitat) [inline src/AggregateFunctions/UniquesHashSet.h:444]: I wouldn't expect reading one value to often be the case and readIntBinary does the same rb.readStrict under the hood (yes, with constant size, but let's neglect this). so, maybe just always use this loop?

File: tests/performance/uniq_stored.xml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            GROUP BY a
+        );
+    </create_query>
+
+    <drop_query>DROP TABLE IF EXISTS mv</drop_query>
+    <drop_query>DROP TABLE IF EXISTS matview</drop_query>
+    <drop_query>DROP TABLE IF EXISTS landing</drop_query>
+
+    <fill_query>
+        insert into landing	select today()-1, toString(rand()%1000),	concat('usr',toString(number%500)) from numbers(10000000);
```
[the remark this answers is not in the document]
Comment (Algunenano) [reply tests/performance/uniq_stored.xml:28]: Makes sense. I'll generate larger dataset (and ideally with less preparation time, which is slow right now).

File: src/AggregateFunctions/UniquesHashSet.h (hunk 1 above)
Comment (Algunenano) [reply src/AggregateFunctions/UniquesHashSet.h:444]: I've tested it and it does have an impact. Not having this makes reading reading statuses with just one value around 25-30% slower. I'll add a benchmark.

Comment (nickitat) [reply src/AggregateFunctions/UniquesHashSet.h:444]: ok

**Review by nickitat — APPROVED** [review]

**Pull request merged**